crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -3041,12 +3029,92 @@ pub(crate) fn check_ast(
     analyze::deferred_scopes(&checker);
 
     let Checker {
-        diagnostics,
-        semantic_errors,
-        ..
+        semantic_errors, ..
     } = checker;
 
-    (diagnostics.into_inner(), semantic_errors.into_inner())
+    semantic_errors.into_inner()
+}
+
+/// A type for collecting diagnostics in a given file.
+///
+/// [`DiagnosticsCollector::report_diagnostic`] can be used to obtain a [`DiagnosticGuard`], which
+/// will push a [`Violation`] to the contained [`OldDiagnostic`] collection on `Drop`.
+pub(crate) struct DiagnosticsCollector<'a> {
+    diagnostics: RefCell<Vec<OldDiagnostic>>,
+    source_file: &'a SourceFile,
+}
+
+impl<'a> DiagnosticsCollector<'a> {
+    /// Create a new collector with the given `source_file` and an empty collection of
+    /// `OldDiagnostic`s.
+    pub(crate) fn new(source_file: &'a SourceFile) -> Self {
+        Self {
+            diagnostics: RefCell::default(),
+            source_file,
+        }
+    }
+
+    /// Return a [`DiagnosticGuard`] for reporting a diagnostic.
+    ///
+    /// The guard derefs to an [`OldDiagnostic`], so it can be used to further modify the diagnostic
+    /// before it is added to the collection in the collector on `Drop`.
+    pub(crate) fn report_diagnostic<'chk, T: Violation>(
+        &'chk self,
+        kind: T,
+        range: TextRange,
+    ) -> DiagnosticGuard<'chk, 'a> {
+        DiagnosticGuard {
+            collector: self,
+            diagnostic: Some(OldDiagnostic::new(kind, range, self.source_file)),
+        }
+    }
+
+    /// Return a [`DiagnosticGuard`] for reporting a diagnostic if the corresponding rule is
+    /// enabled.
+    ///
+    /// Prefer [`DiagnosticsCollector::report_diagnostic`] in general because the conversion from an
+    /// `OldDiagnostic` to a `Rule` is somewhat expensive.
+    pub(crate) fn report_diagnostic_if_enabled<'chk, T: Violation>(
+        &'chk self,
+        kind: T,
+        range: TextRange,
+        settings: &LinterSettings,
+    ) -> Option<DiagnosticGuard<'chk, 'a>> {
+        let diagnostic = OldDiagnostic::new(kind, range, self.source_file);
+        if settings.rules.enabled(diagnostic.rule()) {
+            Some(DiagnosticGuard {
+                collector: self,
+                diagnostic: Some(diagnostic),
+            })
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn into_diagnostics(self) -> Vec<OldDiagnostic> {
+        self.diagnostics.into_inner()
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.diagnostics.borrow().is_empty()
+    }
+
+    pub(crate) fn retain(&self, f: impl FnMut(&OldDiagnostic) -> bool) {
+        self.diagnostics.borrow_mut().retain(f);
```

Comment:
I think this method should take a mut self to avoid mutable borrows (eg by calling collectir.report from inside the callback)